### PR TITLE
fix abcmeta.ABCMeta type

### LINKED ISSUE #5679: abc.ABCMeta.__abstractmethods__ missing on py3+
with:

```python
import abc
import six

@six.add_metaclass(abc.ABCMeta)
class Foo(object):
    @abc.abstractmethod
    def foo(self):
        pass

v = Foo
v.__abstractmethods__
```

`abc.ABCMeta.__abstractmethods__` seems missing on py3+

https://mypy-play.net/?mypy=latest&python=2.7&gist=188750c833a234ed64e29755226a08ab
https://mypy-play.net/?mypy=latest&python=3.9&gist=188750c833a234ed64e29755226a08ab

### PR DESCRIPTION
**Pull request by graingert**
fixes #5679

### CODE REVIEW

File: stdlib/abc.pyi [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
 
 # These definitions have special processing in mypy
 class ABCMeta(type):
-    def register(cls: ABCMeta, subclass: Type[_T]) -> Type[_T]: ...
+    __abstractmethods__: frozenset[str]
+    def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[Any, Any]) -> None: ...
+    def __instancecheck__(cls: ABCMeta, instance: Any) -> Any: ...
+    def __subclasscheck__(cls: ABCMeta, subclass: Any) -> Any: ...
+    def _dump_registry(cls: ABCMeta, file: Optional[SupportsWrite[str]] = None) -> None: ...
+    def register(cls: ABCMeta, subclass: Type[Any]) -> None: ...
```
Comment (graingert) [inline stdlib/abc.pyi:14]:
```suggestion
     def register(cls: ABCMeta, subclass: Type[_T]) -> Type[_T]: ...
```

File: stdlib/abc.pyi [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
 _FuncT = TypeVar("_FuncT", bound=Callable[..., Any])
 
 # These definitions have special processing in mypy
 class ABCMeta(type):
-    def register(cls: ABCMeta, subclass: Type[_T]) -> Type[_T]: ...
+    __abstractmethods__: frozenset[str]
+    def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[Any, Any]) -> None: ...
+    def __instancecheck__(cls: ABCMeta, instance: Any) -> Any: ...
+    def __subclasscheck__(cls: ABCMeta, subclass: Any) -> Any: ...
+    def _dump_registry(cls: ABCMeta, file: Optional[SupportsWrite[str]] = None) -> None: ...
```
Comment (graingert) [inline stdlib/abc.pyi:13]:
```suggestion
    def _dump_registry(cls: ABCMeta, file: Optional[SupportsWrite[str]] = ...) -> None: ...
```

File: stdlib/@python2/abc.pyi [hunk 3]
```diff
@@ -18,7 +18,7 @@ class ABCMeta(type):
     def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[Any, Any]) -> None: ...
     def __instancecheck__(cls: ABCMeta, instance: Any) -> Any: ...
     def __subclasscheck__(cls: ABCMeta, subclass: Any) -> Any: ...
-    def _dump_registry(cls: ABCMeta, *args: Any, **kwargs: Any) -> None: ...
+    def _dump_registry(cls: ABCMeta, file: Optional[SupportsWrite[Any]] = None) -> None: ...
```
Comment (graingert) [inline stdlib/@python2/abc.pyi:21]:
```suggestion
    def _dump_registry(cls: ABCMeta, file: Optional[SupportsWrite[Any]] = ...) -> None: ...
```

File: stdlib/@python2/abc.pyi [hunk 4]
```diff
@@ -18,7 +18,7 @@ class ABCMeta(type):
     def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[Any, Any]) -> None: ...
     def __instancecheck__(cls: ABCMeta, instance: Any) -> Any: ...
     def __subclasscheck__(cls: ABCMeta, subclass: Any) -> Any: ...
-    def _dump_registry(cls: ABCMeta, *args: Any, **kwargs: Any) -> None: ...
+    def _dump_registry(cls: ABCMeta, file: Optional[SupportsWrite[Any]] = ...) -> None: ...
```
Comment (Akuli) [inline stdlib/@python2/abc.pyi:21]: Nit: We now prefer the newer union syntax.
```suggestion
    def _dump_registry(cls: ABCMeta, file: SupportsWrite[Any] | None = ...) -> None: ...
```

Comment (graingert) [reply stdlib/@python2/abc.pyi:21]: I prefer it too, but I copied this from the `print` definition. Do you think it's worth autofixing all the stubs?

File: stdlib/abc.pyi [hunk 5]
```diff
@@ -1,10 +1,16 @@
...
 _T = TypeVar("_T")
 _FuncT = TypeVar("_FuncT", bound=Callable[..., Any])
 
 # These definitions have special processing in mypy
 class ABCMeta(type):
+    __abstractmethods__: frozenset[str]
+    def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[Any, Any]) -> None: ...
+    def __instancecheck__(cls: ABCMeta, instance: Any) -> Any: ...
+    def __subclasscheck__(cls: ABCMeta, subclass: Any) -> Any: ...
+    def _dump_registry(cls: ABCMeta, file: Optional[SupportsWrite[str]] = ...) -> None: ...
```
Comment (graingert) [inline stdlib/abc.pyi:13]:
```suggestion
    def _dump_registry(cls: ABCMeta, file: SupportsWrite[str] | None = ...) -> None: ...
```

File: stdlib/abc.pyi [hunk 6]
```diff
@@ -1,10 +1,16 @@
-from typing import Any, Callable, Type, TypeVar
+from _typeshed import SupportsWrite
+from typing import Any, Callable, Dict, Optional, Tuple, Type, TypeVar
```
Comment (graingert) [inline stdlib/abc.pyi:2]:
```suggestion
from typing import Any, Callable, Dict, Tuple, Type, TypeVar
```

File: stdlib/@python2/abc.pyi [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import _weakrefset
-from typing import Any, Callable, Dict, Set, Tuple, Type, TypeVar
+from _typeshed import SupportsWrite
+from typing import Any, Callable, Dict, Optional, Set, Tuple, Type, TypeVar
```
Comment (graingert) [inline stdlib/@python2/abc.pyi:3]:
```suggestion
from typing import Any, Callable, Dict, Set, Tuple, Type, TypeVar
```

File: stdlib/abc.pyi [hunk 8]
```diff
@@ -1,10 +1,16 @@
...
+from _typeshed import SupportsWrite
+from typing import Any, Callable, Dict, Tuple, Type, TypeVar
 
 _T = TypeVar("_T")
 _FuncT = TypeVar("_FuncT", bound=Callable[..., Any])
 
 # These definitions have special processing in mypy
 class ABCMeta(type):
+    __abstractmethods__: frozenset[str]
+    def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[Any, Any]) -> None: ...
```
Comment (Akuli) [inline stdlib/abc.pyi:10]: I think this is the last nit :) The keys must be strings:

```suggestion
    def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[str, Any]) -> None: ...
```

**Comment by graingert** [comment]
Can I use lowercase Dict and Tuple?

**Comment by Akuli** [comment]
Try it. Sometimes `tuple[Foo, ...]` is buggy, but otherwise it should work.

**Comment by graingert** [comment]
> Try it. Sometimes `tuple[Foo, ...]` is buggy, but otherwise it should work.

sounds like an experiment for another day!

File: stdlib/@python2/abc.pyi [hunk 9]
```diff
@@ -18,7 +18,7 @@ class ABCMeta(type):
     def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[Any, Any]) -> None: ...
```
Comment (graingert) [inline stdlib/@python2/abc.pyi:18]:
```suggestion
    def __init__(self, name: str, bases: Tuple[type, ...], namespace: Dict[str, Any]) -> None: ...
```

Comment (graingert) [reply stdlib/@python2/abc.pyi:18]: @Akuli do these need to be strings also?

**Review by Akuli — APPROVED** [review]

File: stdlib/@python2/abc.pyi (hunk 9 above)
Comment (Akuli) [reply stdlib/@python2/abc.pyi:18]: Honestly I don't care much about Python 2 stubs, but feel free to change it if you do.

**Pull request merged**